baseplate/frameworks/queue_consumer/kafka.py

```diff
@@ -0,0 +1,448 @@
+                    raise ValueError(f"KafkaError: {error.str()}")
+
+                topic = message.topic()
+                offset = message.offset()
+                partition = message.partition()
+
+                span.set_tag("kind", "consumer")
+                span.set_tag("kafka.topic", topic)
+                span.set_tag("kafka.key", message.key())
+                span.set_tag("kafka.partition", partition)
+                span.set_tag("kafka.offset", offset)
+                span.set_tag("kafka.timestamp", message.timestamp())
+
+                blob: bytes = message.value()
+
+                try:
+                    data = self.message_unpack_fn(blob)
+                except Exception:
+                    logger.error("skipping invalid message")
+                    context.trace.incr_tag(f"{self.name}.{topic}.invalid_message")
+                    return
+
+                try:
+                    ingest_timestamp_ms = data["endpoint_timestamp"]
+                    now_ms = int(time.time() * 1000)
+                    message_latency = (now_ms - ingest_timestamp_ms) / 1000
+                except KeyError:
+                    # we can't guarantee that all publishers populate this field
+                    # v2 events publishers (event collectors) do, but future
+                    # kafka publishers may not
+                    message_latency = None
+
+                self.handler_fn(context, data, message)
+
+                if self.on_success_fn:
+                    self.on_success_fn(context, data, message)
+
+                if message_latency is not None:
+                    context.metrics.timer(f"{self.name}.{topic}.latency").send(message_latency)
+
+                context.metrics.gauge(f"{self.name}.{topic}.offset.{partition}").replace(offset)
+        except Exception:
+            # let this exception crash the server so we'll stop processing messages
+            # and won't commit offsets. when the server restarts it will get
+            # this message again and try to process it.
+            logger.exception(
+                "Unhandled error while trying to process a message, terminating the server"
+            )
+            raise
+
+
+class _BaseKafkaQueueConsumerFactory(QueueConsumerFactory):
+    def __init__(
+        self,
+        name: str,
+        baseplate: Baseplate,
+        bootstrap_servers: str,
+        group_id: str,
+        topics: Sequence[str],
+        handler_fn: Handler,
+        message_unpack_fn: Callable[[bytes], Dict[str, Any]] = json.loads,
+        health_check_fn: Optional[HealthcheckCallback] = None,
+    ):
+        self.baseplate = baseplate
+
+        service_name, _, group_name = group_id.partition(".")
+        assert service_name and group_name, "group_id must start with 'SERVICENAME.'"
+        assert name == f"kafka_consumer.{group_name}"
+
+        self.consumer = self._make_consumer(bootstrap_servers, group_id, topics)
```

Comment:
I find that it’s nice to do as little “work” as possible in the constructor and instead provide a builder method that does all the work and just make constructors dumb assignment. 

In the Kafka case, this has been especially useful because systems that have more complicated or specialized needs can set up their own `Queue`s and pass those in rather than relying on baseplate to work for every need.

Reply:
yeah good call